File: 082_remove_duplicates_from_sorted_list_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode curr = head;
        ListNode prev = dummy;
        while (curr != null) {
            ListNode next = curr.next;
            // keep searching for duplicate ones
            int count = 1;
            while (next != null && next.val == curr.val) {
                next = next.next;
                count++;
            }
            if (count > 1) {
                prev.next = next;
            }
            if (count == 1) {
                prev = curr;
            }
            curr = next;
        }

        return dummy.next;
    }
}

File: python/082_remove_duplicates_from_sorted_list_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # Time Complexity: O(N)
    # Space Complexity: O(1)
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        dummy = ListNode(-1)
        dummy.next = head
        prev, curr = dummy, head

        while curr:
            if curr.next and curr.next.val == curr.val:
                val_to_remain = curr.val
                while curr and curr.val == val_to_remain:
                    curr = curr.next
                prev.next = curr
            else:
                prev, curr = curr, curr.next
        return dummy.next
